File: copilot-code/python/src/algorithms/keeper_value.py
"""Keeper value calculation algorithm"""
import logging
from typing import List, Dict, Tuple, Optional
from src.database.player_utils import get_player_position

logger = logging.getLogger(__name__)


def calculate_keeper_value(
    player_id: str,
    keeper_round: int,
    projected_points: float,
    avg_points_at_round: float
) -> float:
    """
    Calculate keeper value with opportunity cost
    
    Args:
        player_id: Player ID
        keeper_round: Round where player can be kept
        projected_points: Player's projected points for season
        avg_points_at_round: Average points of players typically drafted in this round
    
    Returns:
        Keeper value score (higher is better)
    """
    try:
        # Opportunity cost = difference between projected points and typical round value
        opportunity_cost = projected_points - avg_points_at_round
        
        # Bonus for earlier rounds (diminishing returns)
        round_multiplier = 1 + (1 / keeper_round)
        
        keeper_value = opportunity_cost * round_multiplier
        
        logger.debug(f'Keeper value for {player_id}: {keeper_value:.2f}')
        return keeper_value
        
    except Exception as error:
        logger.error(f'Failed to calculate keeper value for {player_id}', exc_info=True)
        return 0.0


def get_best_2_keeper_combo(
    roster: List[Dict],
    max_keepers: int = 2
) -> List[Tuple[str, float]]:
    """
    Find best 2-keeper combination
    
    Args:
        roster: List of eligible keeper candidates with values
        max_keepers: Maximum keepers allowed (default 2)
    
    Returns:
        List of (player_id, value) tuples
    """
    try:
        # Sort by individual keeper value
        sorted_roster = sorted(
            roster,
            key=lambda x: x.get('keeper_value', 0),
            reverse=True
        )
        
        # Return top N keepers
        best_keepers = [
            (player['player_id'], player['keeper_value'])
            for player in sorted_roster[:max_keepers]
        ]
        
        logger.info(f'Best {max_keepers} keeper combo: {best_keepers}')
        return best_keepers
        
    except Exception as error:
        logger.error('Failed to calculate best keeper combo', exc_info=True)
        return []


def get_projected_points(player_id: str, season: int) -> float:
    """
    Get projected points for a player
    TODO: Implement actual projection logic based on historical performance
    
    Args:
        player_id: Player ID
        season: Target season
    
    Returns:
        Projected fantasy points
    """
    # TODO: Calculate from database statistics
    return 0.0


def get_average_value_at_round(round_num: int, position: Optional[str] = None) -> float:
    """
    Get average fantasy points of players typically drafted in this round
    TODO: Implement based on historical draft data
    
    Args:
        round_num: Draft round number
        position: Optional position filter
    
    Returns:
        Average points value
    """
    # TODO: Calculate from database
    return 0.0

File: copilot-code/typescript/src/algorithms/keeper-value.ts
import { logger } from '../config';
import { KeeperPrediction } from '../types';

/**
 * Calculate keeper value for a player
 * 
 * Formula:
 * baseValue = projectedRound - keepCost
 * opportunityCost = player.projectedPoints - avgPlayerValueAtRound(keepCost)
 * value = (baseValue * ownerMultiplier * scarcityMultiplier) + (opportunityCost * 0.3)
 * 
 * Weights:
 * - Owner history: Primary weight
 * - Position scarcity: Secondary weight
 * - Opportunity cost: 30% weight
 */

interface KeeperValueInput {
  playerId: string;
  playerName: string;
  position: string;
  projectedRound: number;
  projectedPoints: number;
  keepCost: number;
  ownerMultiplier: number; // From owner's keeper history
  scarcityMultiplier: number; // From position scarcity analysis
  avgValueAtRound: number;
}

export function calculateKeeperValue(input: KeeperValueInput): number {
  try {
    const {
      projectedRound,
      keepCost,
      projectedPoints,
      ownerMultiplier,
      scarcityMultiplier,
      avgValueAtRound,
    } = input;

    // Base value: difference between projected round and keeper cost
    const baseValue = projectedRound - keepCost;

    // Opportunity cost: what are we giving up by keeping this player?
    const opportunityCost = projectedPoints - avgValueAtRound;

    // Final value calculation
    const value =
      baseValue * ownerMultiplier * scarcityMultiplier +
      opportunityCost * 0.3;

    logger.debug('Keeper value calculated', {
      playerId: input.playerId,
      baseValue,
      opportunityCost,
      finalValue: value,
    });

    return value;
  } catch (error) {
    logger.error('Failed to calculate keeper value', {
      playerId: input.playerId,
      error: error instanceof Error ? error.message : String(error),
    });
    throw error;
  }
}

/**
 * Find the best 2-keeper combination for a team
 * Considers both individual values and position diversity
 */
export function getBest2KeeperCombo(
  players: KeeperValueInput[],
  rosterNeeds: { [position: string]: number }
): {
  players: [KeeperValueInput, KeeperValueInput];
  totalValue: number;
  positionDiversity: number;
} | null {
  try {
    if (players.length < 2) {
      logger.warn('Not enough players for 2-keeper combo');
      return null;
    }

    let bestCombo: {
      players: [KeeperValueInput, KeeperValueInput];
      totalValue: number;
      positionDiversity: number;
    } | null = null;

    // Generate all combinations of 2 players
    for (let i = 0; i < players.length; i++) {
      for (let j = i + 1; j < players.length; j++) {
        const player1 = players[i];
        const player2 = players[j];

        const value1 = calculateKeeperValue(player1);
        const value2 = calculateKeeperValue(player2);
        const totalValue = value1 + value2;

        // Position diversity bonus (20% if different positions)
        const positionDiversity =
          player1.position !== player2.position ? 1.2 : 1.0;

        const adjustedValue = totalValue * positionDiversity;

        if (!bestCombo || adjustedValue > bestCombo.totalValue * bestCombo.positionDiversity) {
          bestCombo = {
            players: [player1, player2],
            totalValue,
            positionDiversity,
          };
        }
      }
    }

    logger.info('Best 2-keeper combo calculated', {
      totalValue: bestCombo?.totalValue,
      positionDiversity: bestCombo?.positionDiversity,
    });

    return bestCombo;
  } catch (error) {
    logger.error('Failed to calculate best 2-keeper combo', {
      error: error instanceof Error ? error.message : String(error),
    });
    throw error;
  }
}

/**
 * Calculate average player value at a specific round
 * TODO: Query database for historical average points by round
 */
export async function getAverageValueAtRound(round: number): Promise<number> {
  try {
    // TODO: Implement database query
    // SELECT AVG(projected_points) FROM players
    // JOIN draft_picks ON players.player_id = draft_picks.player_id
    // WHERE draft_picks.round = $1

    logger.debug('Getting average value at round', { round });

    // Placeholder: Decreasing value by round
    const baseValue = 200;
    const decayRate = 10;
    const avgValue = Math.max(0, baseValue - round * decayRate);

    return avgValue;
  } catch (error) {
    logger.error('Failed to get average value at round', {
      round,
      error: error instanceof Error ? error.message : String(error),
    });
    throw error;
  }
}

/**
 * Get owner's keeper pattern multiplier
 * TODO: Query database for owner's historical keeper tendencies
 */
export async function getOwnerMultiplier(
  userId: string,
  position: string
): Promise<number> {
  try {
    // TODO: Implement database query
    // SELECT frequency FROM manager_tendencies
    // WHERE user_id = $1 AND position = $2
    // AND round <= 5 -- Focus on early keeper decisions

    logger.debug('Getting owner multiplier', { userId, position });

    // Placeholder: Default to 1.0
    return 1.0;
  } catch (error) {
    logger.error('Failed to get owner multiplier', {
      userId,
      position,
      error: error instanceof Error ? error.message : String(error),
    });
    throw error;
  }
}

/**
 * Get position scarcity multiplier
 * TODO: Query database for position scarcity analysis
 */
export async function getScarcityMultiplier(position: string): Promise<number> {
  try {
    // TODO: Implement database query
    // Calculate based on:
    // 1. Number of startable players at position
    // 2. Drop-off in value after elite tier
    // 3. Total draft picks at position historically

    logger.debug('Getting scarcity multiplier', { position });

    // Placeholder: Position-based scarcity
    const scarcityMap: { [pos: string]: number } = {
      QB: 0.9, // Less scarce
      RB: 1.3, // Very scarce
      WR: 1.1, // Moderately scarce
      TE: 1.2, // Scarce after elite tier
      K: 0.8,
      DEF: 0.8,
    };

    return scarcityMap[position] || 1.0;
  } catch (error) {
    logger.error('Failed to get scarcity multiplier', {
      position,
      error: error instanceof Error ? error.message : String(error),
    });
    throw error;
  }
}
